fluent-bit/fluent-bit-go-loki/out_loki.go

```diff
@@ -0,0 +1,197 @@
+	return output.GetRecord(dec)
+}
+
+func (p *fluentPlugin) NewDecoder(data unsafe.Pointer, length int) *output.FLBDecoder {
+	return output.NewDecoder(data, length)
+}
+
+func (p *fluentPlugin) Exit(code int) {
+	os.Exit(code)
+}
+
+func (p *fluentPlugin) HandleLine(ls model.LabelSet, timestamp time.Time, line string) error {
+	return loki.Handle(ls, timestamp, line)
+}
+
+//export FLBPluginRegister
+func FLBPluginRegister(ctx unsafe.Pointer) int {
+	return output.FLBPluginRegister(ctx, "loki", "Ship fleunt-bit logs to Grafana Loki")
+}
+
+//export FLBPluginInit
+// (fluentbit will call this)
+// ctx (context) pointer to fluentbit context (state/ c code)
+func FLBPluginInit(ctx unsafe.Pointer) int {
+	// Example to retrieve an optional configuration parameter
+	url := plugin.PluginConfigKey(ctx, "URL")
+	batchWait := plugin.PluginConfigKey(ctx, "BatchWait")
+	batchSize := plugin.PluginConfigKey(ctx, "BatchSize")
+	labels := plugin.PluginConfigKey(ctx, "Labels")
+	logLevel := plugin.PluginConfigKey(ctx, "LogLevel")
+	removeKeyStr := plugin.PluginConfigKey(ctx, "RemoveKeys")
+
+	config, err := getLokiConfig(url, batchWait, batchSize, labels, logLevel, removeKeyStr)
+	if err != nil {
+		level.Error(logger).Log("[flb-go]", "failed to launch", "error", err)
+		plugin.Unregister(ctx)
+		plugin.Exit(1)
+		return output.FLB_ERROR
+	}
+	logger = newLogger(config.logLevel)
+	level.Info(logger).Log("[flb-go]", "Starting fluent-bit-go-loki", "version", version.Info())
+	level.Info(logger).Log("[flb-go]", "provided parameter", "URL", url)
+	level.Info(logger).Log("[flb-go]", "provided parameter", "BatchWait", batchWait)
+	level.Info(logger).Log("[flb-go]", "provided parameter", "BatchSize", batchSize)
+	level.Info(logger).Log("[flb-go]", "provided parameter", "Labels", labels)
+	level.Info(logger).Log("[flb-go]", "provided parameter", "LogLevel", logLevel)
+	level.Info(logger).Log("[flb-go]", "provided parameter", "RemoveKeys", removeKeyStr)
+
+	cfg := client.Config{}
+	// Init everything with default values.
+	flagext.RegisterFlags(&cfg)
+
+	// Override some of those defaults
+	cfg.URL = config.url
+	cfg.BatchWait = config.batchWait
+	cfg.BatchSize = config.batchSize
+
+	log := logrus.New()
+
+	loki, err = client.New(cfg, kit.NewLogrusLogger(log))
+	if err != nil {
+		level.Error(logger).Log("client.New", err)
+		plugin.Unregister(ctx)
+		plugin.Exit(1)
+		return output.FLB_ERROR
+	}
+	ls = config.labelSet
+	removeKeys = config.removeKeys
+
+	return output.FLB_OK
+}
+
+//export FLBPluginFlush
+func FLBPluginFlush(data unsafe.Pointer, length C.int, tag *C.char) int {
+	var ret int
+	var ts interface{}
+	var record map[interface{}]interface{}
+
+	dec := plugin.NewDecoder(data, int(length))
+
+	for {
+		ret, ts, record = plugin.GetRecord(dec)
+		if ret != 0 {
+			break
+		}
+
+		// Get timestamp
+		var timestamp time.Time
+		switch t := ts.(type) {
+		case output.FLBTime:
+			timestamp = ts.(output.FLBTime).Time
+		case uint64:
```

Comment:
is this case required ? How do you know this can happens ?

Reply:
This should be required because Fluentd events can contain two time formats which are `EventTime` and Unix epoch.

`EventTime` will be translated as `output.FLBTime` type.
Unix epoch will be translated as uint64 type.

ref: https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#message-modes
ref: https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#eventtime-ext-format